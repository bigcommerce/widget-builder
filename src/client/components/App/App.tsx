import React, { Component, createRef } from 'react';
import { io, Socket } from 'socket.io-client';
import {
    Flex,
    GlobalStyles,
    H1,
} from '@bigcommerce/big-design';

import { host, port } from '../../../config';
import { createElementFromHTML, executeWidgetScripts } from '../../utils/widget';
import { SocketData } from '../../../types';
import { loadSlick } from '../../utils/slick';

export class App extends Component {
    private widgetFrameRef: React.RefObject<HTMLDivElement> = createRef<HTMLDivElement>();

    socket?: Socket;

    componentDidMount() {
        this.socket = io(`${host}:${port}`);

        this.socket.on('connect', () => console.log('Socket connected')); // eslint-disable-line no-console
        this.socket.on('event', (data: SocketData) => {
            this.mountWidget(data.html);
        });
        this.socket.on('disconnect', () => {
            this.socket = undefined;
        });
    }

    mountWidget(html: string) {
        const widgetFrame = this.widgetFrameRef.current as HTMLDivElement;

        while (widgetFrame.firstChild) {
            widgetFrame.removeChild(widgetFrame.firstChild);
        }
        const widgetElement = createElementFromHTML(html);
        widgetFrame.appendChild(widgetElement);
        executeWidgetScripts(widgetElement, document);
        loadSlick();
    }

    render() {
        return (
            <div>
                <GlobalStyles/>
                <Flex
                    alignContent={'center'}
                    flexDirection={'row'}
                    justifyContent={'center'}
                    padding={'medium'}
                >
                    <img
                        src={'favicon.ico'}
                        height={34}
                    />
                    &nbsp;&nbsp;&nbsp;
                    <H1>Widget Builder</H1>
                </Flex>
                <div ref={this.widgetFrameRef}></div>
            </div>
        );
    }
}

export default App;
