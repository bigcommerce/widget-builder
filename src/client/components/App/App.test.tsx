import React from 'react';
import { mount } from 'enzyme';

import { App } from './App';

jest.mock('socket.io-client', () => ({
    io: () => ({
        on: () => ({}),
    }),
}));

it('renders', () => {
    const app = mount(<App/>);
    expect(app.getDOMNode()).toMatchSnapshot();
});

describe('mountWidget', () => {
    it('correctly mounts a widget', () => {
        const app = mount(<App/>);
        (app.instance() as App).mountWidget('<div>blah</div>');
        expect(app.getDOMNode()).toMatchSnapshot();
    });
});
