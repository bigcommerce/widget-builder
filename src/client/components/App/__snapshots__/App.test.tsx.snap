// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mountWidget correctly mounts a widget 1`] = `
<div>
  <div
    class="styled__StyledBox-sc-sj5f1m-0 VUKyh styled__StyledFlex-sc-hcvk8l-0 iBHbGA"
  >
    <img
      height="34"
      src="favicon.ico"
    />
       
    <h1
      class="styled__StyledH1-sc-tqnj75-1 klUKKw"
    >
      Widget Builder
    </h1>
  </div>
  <div>
    <div>
      blah
    </div>
  </div>
</div>
`;

exports[`renders 1`] = `
<div>
  <div
    class="styled__StyledBox-sc-sj5f1m-0 VUKyh styled__StyledFlex-sc-hcvk8l-0 iBHbGA"
  >
    <img
      height="34"
      src="favicon.ico"
    />
       
    <h1
      class="styled__StyledH1-sc-tqnj75-1 klUKKw"
    >
      Widget Builder
    </h1>
  </div>
  <div />
</div>
`;
